spring-web/src/main/java/org/springframework/http/client/JdkClientHttpRequest.java

```diff
@@ -92,28 +94,46 @@ public URI getURI() {
 	@Override
 	@SuppressWarnings("NullAway")
 	protected ClientHttpResponse executeInternal(HttpHeaders headers, @Nullable Body body) throws IOException {
+		HttpRequest request = buildRequest(headers, body);
+		CompletableFuture<HttpResponse<InputStream>> responsefuture =
+				this.httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofInputStream());
 		try {
-			HttpRequest request = buildRequest(headers, body);
-			HttpResponse<InputStream> response;
 			if (this.timeout != null) {
-				response = this.httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofInputStream())
-						.get(this.timeout.toMillis(), TimeUnit.MILLISECONDS);
-			}
-			else {
-				response = this.httpClient.send(request, HttpResponse.BodyHandlers.ofInputStream());
+				CompletableFuture<Void> timeoutFuture = new CompletableFuture<Void>()
+						.completeOnTimeout(null, this.timeout.toMillis(), TimeUnit.MILLISECONDS);
+				timeoutFuture.thenRun(() -> {
+					if (!responsefuture.cancel(true) && !responsefuture.isCompletedExceptionally()) {
+						try {
+							responsefuture.resultNow().body().close();
```

Comment:
Can we really be sure there is a body, I imagine the timeout could happen before the server responds with status and headers.